crates/red_knot_python_semantic/src/types/call/arguments.rs

```diff
@@ -1,88 +1,179 @@
+use std::cell::Cell;
+use std::rc::Rc;
+
+use bitflags::bitflags;
+
 use super::Type;
 
-/// Typed arguments for a single call, in source order.
+/// Typed arguments for a single call, in source order, with an optional bound self/cls parameter.
 #[derive(Clone, Debug, Default)]
-pub(crate) struct CallArguments<'a, 'db>(Vec<Argument<'a, 'db>>);
+pub(crate) struct CallArguments<'a, 'db> {
+    bound_self: Option<Argument<'a, 'db>>,
+    arguments: Rc<[Argument<'a, 'db>]>,
+}
 
 impl<'a, 'db> CallArguments<'a, 'db> {
     /// Create a [`CallArguments`] with no arguments.
     pub(crate) fn none() -> Self {
-        Self(Vec::new())
+        Self {
+            bound_self: None,
+            arguments: vec![].into(),
```

Comment:
Nit: Would `Default::default()` work?
```suggestion
            arguments: Default::default(),
```

Reply:
Done (now with `VecDeque`)